mgrachev/same tree/Program.cs

```diff
@@ -0,0 +1,126 @@
+﻿/*
+Given the roots of two binary trees p and q, write a function to check if they are the same or not.
+Two binary trees are considered the same if they are structurally identical, and the nodes have the same value.
+
+ Example 1:
+
+   1      1
+  / \    / \
+ 2   3  2   3
+
+Input: p = [1,2,3], q = [1,2,3]
+Output: true
+
+Example 2:
+
+   1   1
+  /     \
+ 2       2
+
+Input: p = [1,2], q = [1,null,2]
+Output: false
+
+Example 3:
+   1      1
+  / \    / \
+ 2   1  1   2
+
+Input: p = [1,2,1], q = [1,1,2]
+Output: false 
+
+Constraints:
+The number of nodes in both trees is in the range [0, 100].
+-104 <= Node.val <= 104
+*/
+
+#pragma warning disable 8604
+using System.Diagnostics;
+
+// array of tuples of TreeNode
+var testCases = new(TreeNode, TreeNode, bool) [] 
+{
+    (
+        new TreeNode(1,new TreeNode(2),new TreeNode(3)),
+        new TreeNode(1,new TreeNode(2),new TreeNode(3)),
+        true
+    ),
+    (
+        new TreeNode(1,new TreeNode(2),null),
+        new TreeNode(1,null,new TreeNode(2)),
+        false
+    ),
+    (
+        new TreeNode(1,new TreeNode(2),new TreeNode(1)),
+        new TreeNode(1,new TreeNode(1),new TreeNode(2)),
+        false
+    )
+};
+
+var sol = new Solution();
+var i = 0;
+foreach(var tc in testCases)
+{
+    Console.WriteLine($"Test {++i} (expected {tc.Item3})");
+    Console.WriteLine($"Recursive: {sol.IsSameTreeRecursive(tc.Item1,tc.Item2)}");
+    Console.WriteLine($"Iterative: {sol.IsSameTreeIterative(tc.Item1,tc.Item2)}\n");
+};
+
+// Definition for a binary tree node.
+public class TreeNode {
+    public int Val {get;set;}
+    public TreeNode Left {get;set;}
+    public TreeNode Right {get;set;}
+
+    public TreeNode(int val=0) {
+        this.Val = val;
+    }
+    public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
+        this.Val = val;
+        this.Left = left;
+        this.Right = right;
+    }
+}
+
+public class Solution {
+    public bool IsSameTreeRecursive(TreeNode p, TreeNode q) {
+        if(p==null || q==null)
+        {
+            return (p == null && q==null);
+        }
+        if (p.Val!=q.Val) return false;
+         
+        return IsSameTreeRecursive(p.Left, q.Left) && IsSameTreeRecursive(p.Right, q.Right);
+    }
+    
+    public bool IsSameTreeIterative(TreeNode? p, TreeNode? q) {
```

Comment:
Thanks for bringing in iterative solution, I was starting to implement in this way myself but it did not work with [1,2] and [1, null, 2] example, comparison sometimes is done "left with right". Did not have time to figure it out

my version was like this:

bool IsSameTree(TreeNode p, TreeNode q)
        {
            if (q == null && p == null)
            {
                return true;
            }

            if ((p == null && q != null) || (q == null && p != null))
            {
                return false;
            }


            Stack<TreeNode> pStack = new Stack<TreeNode>();
            pStack.Push(p);
            Stack<TreeNode> qStack = new Stack<TreeNode>();
            qStack.Push(q);

            while (pStack.Any() && qStack.Any())
            {
                TreeNode pNode = pStack.Pop();
                TreeNode qNode = qStack.Pop();

                if (pNode.val != qNode.val)
                {
                    return false;
                }

                if (pNode.right != null)
                {
                    pStack.Push(pNode.right);
                }

                if (qNode.right != null)
                {
                    qStack.Push(qNode.right);
                }

                if (pNode.left != null)
                {
                    pStack.Push(pNode.left);
                }

                if (qNode.left != null)
                {
                    qStack.Push(qNode.left);
                }
            }

            // both empty meaning same length
            if (!pStack.Any() && !qStack.Any()) 
            {
                return true;
            }

            return false;
            
        }